### DISCUSSION TITLE
Are only 8-bit channel image formats supported?

### DISCUSSION THREAD

**Discussion by DevChagrins** [zxing-cpp/zxing-cpp#537] [Q&A] [answered]
I've been working on a project using the ZXing-CPP library and from looking over the code, it seems like it extracts a pixel in a series of 3 or 4 8-bit channels. I just wanted to confirm if what I was reading is true and there isn't any support for any larger or smaller channel formats without making a custom solution.

**Comment by axxel** [comment, marked as answer]
You generally got it right but:
1) besides 3 and 4 there is also direct support for 1 8-bit channel.
2) higher bit-depth input might also work if you can live with masking or slicing 8 bits from them. Say you have a 16-bit grey-scale image (pointed to by `uint16_t* data`, you can then create an `ImageView` object that uses the higher 8-bits (on a little endian system) like so:
```
auto iv = ImageView(static_cast<uint8_t*>(data) + 1, width, height, ImageFormat::Lum, width * 2, 2);
```

If you have a specific pixel format and wonder how that might be used, let me know which one it is.

**Reply by DevChagrins** [reply]
I'll let you know as I progress in the project. I'm not sure how various hardware platforms will return camera data exactly, but thanks for the answer!

**Reply by axxel** [reply]
If you are dealing with raw bits from cameras you might want to have a look at the [qtwrapper](https://github.com/zxing-cpp/zxing-cpp/blob/5634526c2b7307d78e3ce477e1adfd61c362d9d7/example/ZXingQtReader.h#L195) for inspiration. Most important takeaway: if you have some YUV-style input, simply use the 'Y' channel as `ImageFormat::Lum` and don't first convert that to RGB, which is a complete waste of cpu instructions.
